### PR TITLE
docs: Add section on the OPA document model

### LINKED ISSUE #2284: Documentation on base/virtual document lost in upgrade
# Expected Behavior

The docs used to have a section on the base document, virtual documents and their relation to policies: https://www.openpolicyagent.org/docs/v0.13.5/how-does-opa-work/

These docs were very useful to help connect the OPA dots!

# Actual Behavior

No current documentation covering the topic.

# Additional Info

There are still links scattered around in the docs pointing to the "how-does-opa-work" page for en explanation of the base/virtual document, though that's not really explained by that page anymore.

### PR DESCRIPTION
**Pull request by tsandall**
When we refreshed the docs last year we removed the page titled 'how
OPA works' and in the process lost the description of the OPA document
model that explains the base and virtual document concepts. Since then
a bunch of people have referred back to those docs after getting
started.

This change re-introduces the model and provides an explanation of
_why_ the model exists.

Fixes #2284

Signed-off-by: Torin Sandall <torinsandall@gmail.com>

### CODE REVIEW

File: docs/content/philosophy.md [hunk 1]
```diff
@@ -78,4 +78,77 @@ dynamically into OPA remotely via APIs or through the local filesystem.
...
 OPA is a full-featured policy engine that offloads policy decisions from your
 service. You can think of it as a concierge for your service who can answer
-detailed questions on behalf of your users to meet their specific needs.
\ No newline at end of file
+detailed questions on behalf of your users to meet their specific needs.
+
+## How Does OPA Work? {#how-does-opa-work}
+
+> 📖 This section assumes you are familiar with the basics of OPA. We recommend
+> going through the [Introduction](..) before reading this section.
```
Comment (timothyhinrichs) [inline docs/content/philosophy.md:86]: The requirement for going through the intro doesn't seem like the right thing for this page.  If you're reading this page you  you don't know what OPA is (section 1) or why you might use it (section 2).

File: docs/content/philosophy.md [hunk 2]
```diff
@@ -78,4 +78,77 @@ dynamically into OPA remotely via APIs or through the local filesystem.
...
+
+OPA policies (written in Rego) make decisions based on hierarchical structured data.
```
Comment (timothyhinrichs) [inline docs/content/philosophy.md:88]: I'd be tempted to copy in the usual diagram and explanation from the Intro here.  Maybe as its own section.  That's "How OPA Works".  Then maybe this section is something titled maybe "OPA's information model" (though "information model" bothers me).

File: docs/content/philosophy.md [hunk 3]
```diff
@@ -78,4 +78,77 @@ dynamically into OPA remotely via APIs or through the local filesystem.
...
+
+The following table summarizes how base documents (loaded using different models)
+can be referenced by your policies and integrated into OPA.
+
+| Model | How to access in Rego | How to integrate with OPA |
+| --- | --- | --- |
+| Synchronous Push | The `input` global variable | Provide data in policy query (e.g., `POST v1/data`) |
+| Synchronous Pull | The built-in functions in OPA (e.g., `http.send`) | Provide an API that rules can query. |
+| Asynchronous Push | The `data` global variable | Invoke OPA API(s) (e.g., `PUT v1/data`)
+| Asynchronous Pull | The `data` global variable | Configure OPA's [Bundle](../management#bundles) feature. |
```
Comment (timothyhinrichs) [inline docs/content/philosophy.md:117]: This table feels too detailed.  I was scratching my head for a while about the Model column.  Before this table, where I  it was going was to say that all of the base and virtual documents are accessible through the API and through Rego from a single namespace: `data`.    And then continue on.  

The discussion of in-memory cache, etc. is good but feels that it should come later. once people understand how to think about /data.    Something like...
* defining "virtual docs" and "base docs" as above
* virtual/base docs unified via data.  API and Rego expressions.
* Controlling where virtual docs exist in the hierarchy via `package`
* Controlling where base docs exist in the hiearchy and the different options for loading data (e.g. this table)

File: docs/content/philosophy.md (hunk 1 above)
Comment (tsandall) [reply docs/content/philosophy.md]: Agreed. I've amended the PR and I'm happier with the new version.

File: docs/content/philosophy.md (hunk 2 above)
Comment (tsandall) [reply docs/content/philosophy.md:96]: Yep. What happened was that I couldn't find a good title. I didn't like the way the different options looked in the ToC, lol. I ended up settling on How OPA Works only to realize we needed to refer to the standard overview stuff.

Anyway, I think it's better in the latest version. I've put a link to the introduction that contains the standard overview and I'm fine with the new title ("The OPA Document Model"). Does that title work for you?

File: docs/content/philosophy.md (hunk 3 above)
Comment (tsandall) [reply docs/content/philosophy.md]: Thanks! This was super helpful. I moved the table to the end and fleshed out a few things that I thought were worthwhile:

* Mention the "arbitrary data" concept right off the bat
* Better discussion of naming/locations. This lets us answer why `input` exists for example.

File: docs/content/philosophy.md [hunk 4]
```diff
@@ -74,8 +74,104 @@ dynamically into OPA remotely via APIs or through the local filesystem.
...
+called `data`. Similarly, OPA lets you query for both base and virtual documents via the
+`/v1/data` HTTP API [3]. This is why queries for just `data` (or `data.foo` or `data.foo.bar`, etc.)
+return the combination of base and virtual documents located under that path.
+
+Since base documents come from outside of OPA, their location under `data` is controlled
+by the software doing the loading. On the other hand, the location of virtual
+documents under `data` is controlled by policies themselves using the `package` directive
+in the language.
+
+Sometimes base documents need to be loaded into OPA synchronously with respect to policy
```
[the remark this answers is not in the document]
Comment (timothyhinrichs) [reply docs/content/philosophy.md:130]: I'd be tempted to have 3 choices: synchronous -> input, asynchronous -> data, dynamic (http.send) -> local variables, possibly under data.

File: docs/content/philosophy.md (hunk 2 above)
Comment (timothyhinrichs) [reply docs/content/philosophy.md:96]: Title seems good.

**Pull request merged**